Api compat changes

weshaggard opened the pull request:
Fixes https://github.com/Azure/azure-sdk-for-net/issues/12546

PTAL @pakrym @chidozieononiwu

weshaggard committed:
Update how ApiCompat checks are ran

- Now include the ApiCompatVersion in the project
- Update the ApiCompatVersion everytime we ship a GA
- Remove EnableApiCompat as to disable now remove the ApiCompatVersion

weshaggard committed:
Remove EnableApiCompat from everywhere.

weshaggard committed:
Add ApiCompatVersion to all GA packages

drwill-ms commented on sdk/digitaltwins/Azure.DigitalTwins.Core/src/Azure.DigitalTwins.Core.csproj:
```diff
@@ -4,7 +4,6 @@
     <AssemblyTitle>Digital Twins SDK</AssemblyTitle>
     <TargetFrameworks>$(RequiredTargetFrameworks)</TargetFrameworks>
 
-    <EnableApiCompat>false</EnableApiCompat>
```
Is removal alone sufficient? We haven't done the post-release work yet, but it is on our task list.

weshaggard replied:
Yes with my update to how ApiCompat works you won't get Api checks done until you add an ApiCompatVersion in your project. That will happen automatically as part of our version bump PR once you ship a GA.

drwill-ms replied:
Do we have to ship a GA, or can our shipped public preview work for getting this task done?

weshaggard replied:
You can add any shipped version if you want to start a baseline to have the API's verified againt it.

JoshLove-msft replied:
I wonder if there really is a use case where you would want to specify some version other than what the version ApiCompatVersion gets auto-updated to, or is storing off the version number in the project just a technical necessity? I guess what I'm getting at is would it be possible to have this value not be configurable? Turning off or on could be configurable but selecting a version would not be. I suppose this might not be possible since we would need to store off the version to do the compat check separately from the actual package version to account for previews, but my worry is that this might get manually updated to an incorrect value. If it was in a shared props file outside of the package, it might make this less likely and also avoid potential confusion around what the value should be set to.

weshaggard replied:
As you point out we would have to keep it somewhere and personally I think keeping it right next the current version is the best spot for it. Sure someone could change it but they could change it no matter were we put it. The good news is that our automation for bumping the version will always try to update it once we ship a GA so it will eventually get reset back.

As for specifically setting it to something else I can imagine cases like digitaltwins that wants to lock the api surface while still in preview and so they could set it to a earlier preview version and then ensure no breaking changes happen after that.

JoshLove-msft replied:
I think the most likely situation where someone might incorrectly change this property would be when doing a Preview version for a GA library, as this is the only time we need to manually change the version. Since the property is right next to the package version, they might think that they need to update both. However, I think the build would fail if they do this since that version hasn't been shipped yet, so I would agree that the potential for harm is pretty low. BTW this is awesome that we no longer have to remember to update this when shipping a new GA version.

maririos commented:
@weshaggard Could you please add `<ApiCompatVersion>1.0.0</ApiCompatVersion>` to Text Analytics? We released GA yesterday

heaths reviewed (approved):
Key Vault and Search changes LGTM. Thanks!

weshaggard commented:
>@weshaggard Wes Haggard FTE Could you please add <ApiCompatVersion>1.0.0</ApiCompatVersion> to Text Analytics? We released GA yesterday 

Will do.

weshaggard committed:
PR feedback

Fix bad appconfig test
Add ApiCompatVersion for textanalytics

maririos reviewed (approved):
Text Analytics and Form Recognizer LGTM.
Thanks Wes!

jsquire reviewed (approved):
Event Hubs bits look good to me.

christothes reviewed (approved):
👍 for tables

pakrym commented on eng/ApiCompat/ApiCompat.csproj:
```diff
@@ -1,41 +1,31 @@
   <ItemGroup>
-    <PackageReference Include="Microsoft.DotNet.ApiCompat" />
+    <PackageReference Include="Microsoft.DotNet.ApiCompat" Version="5.0.0-beta.19552.1" />
+    <PackageReference Include="$(TargetPackageName)" Version="$(TargetPackageVersion)" />
```
How does this work in VS when you have multiple projects open?

weshaggard replied:
This project should never really be loaded in VS but you bring up a good question that I should verify that these changes work in VS which I haven't done yet.

pakrym replied:
We had problems with building other projects when this one is not loaded in VS.

Currently it's added to all solutions.

pakrym replied:
Also, does this work well without requiring a manual restore?

weshaggard replied:
The manual restore is not at all needed with this solution, as we restore as part of the call to the run the apicompat target. We no longer have a projectreference to this project.

weshaggard replied:
OK I just did some basic testing in VS (used keyvault solution for testing) and while the APICompat project is in the solution it will build but will not end up doing anything interesting when building without the extra properties, it doesn't fail but it also doesn't do anything interesting. The ApiCompat checks still work as expected for the other projects when building in VS.

So in the end things appear to work as I expect and we should just remove the ApiCompat project from the solutions. I'll make a pass over those and clean those up as part of this PR.

pakrym replied:
Last question, how's the build time considering we do additional restore per every project?

weshaggard replied:
I haven't looked at before and after build times but I don't think it will be much of an impact because the Restore step should be a no-op after the first time. I will take a look at a core build to see if there is any notice impact.

weshaggard replied:
I looked at a few random builds before and after my change of the core pipeline and the build and package are both in the 7min range so nothing really jumps out. However with that said please do let me know if anyone sees any noticeable performance issues.

christothes commented:
[This](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/keyvault/Azure.Security.KeyVault.Administration/src/Azure.Security.KeyVault.Administration.csproj#L9) just merged this morning. Would you mind removing the EnableAPpiompat property?

weshaggard committed:
Remove apicompat project from solutions

weshaggard committed:
Remove more EnableApiCompat properties

weshaggard commented:
@christothes I updated my PR and removed the EnableApiCompat property from your project but now it looks like my CI failures are caused by your project. Are there some api generation missing? Can you please take a look.
